### PR TITLE
s3: generic upload component

### PR DESCRIPTION
**Pull request by jamesacklin**
Adds S3 upload, cribbing a lot of work from https://github.com/arthyn/silo.

Introduces an otherwise unremarkable HTML file input component which accepts a forwarded ref. The thinking here is that we could use this elsewhere in the interface as a hidden element, pass the input an event the ref, and use the provided FileStore however we see fit.

### CODE REVIEW

**Comment by liam-fitzgerald** [comment]
pls watch the `package-lock` thrashing!

File: ui/src/components/FileUpload.tsx [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      },
+      [fs, s3]
+    );
+
+    useEffect(() => {
+      files.forEach((upload: Upload, idx: number) => {
+        uploadFile(upload, idx);
+      });
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [files.length]);
```
Comment (liam-fitzgerald) [inline ui/src/components/FileUpload.tsx:78]: accessors don't work properly in dependencies afaik?

File: ui/src/components/FileUpload.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+              new PutObjectCommand({
+                Bucket: s3.configuration.currentBucket,
+                Key: `${window.ship}/${timestamp}-${file.name}`,
+                Body: file,
+                ContentType: file.type,
+                ContentLength: file.size,
+                ACL: 'public-read',
+              })
+            )
+            .then(() => {
```
Comment (liam-fitzgerald) [inline ui/src/components/FileUpload.tsx:58]: please don't mix `await` and `.then()` unless absolutely necessary

File: ui/src/components/FileUpload.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    useEffect(() => {
+      const hasCredentials =
+        credentials?.accessKeyId &&
+        credentials?.endpoint &&
+        credentials?.secretAccessKey;
+      if (hasCredentials) {
+        const connect = async () => {
+          const client = createClient(credentials);
+          return client;
```
Comment (liam-fitzgerald) [inline ui/src/components/FileUpload.tsx:30]: if you return a promise directly, the calling function doesn't need to be `async`, the result is still awaitable though

File: ui/src/components/FileUpload.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      },
+      [fs]
+    );
+
+    if (storage.loaded)
+      return (
+        <input
+          type="file"
+          multiple={multiple}
+          onChange={(e) => onFiles(e)}
```
Comment (liam-fitzgerald) [inline ui/src/components/FileUpload.tsx:97]: drop unnecessary arrow function

**Review by liam-fitzgerald — CHANGES_REQUESTED** [review]
looks mostly good, but some of the async work needs fixing

File: ui/src/components/FileUpload.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+const UploadInput = React.forwardRef<HTMLInputElement, UploadInputProps>(
+  (props, ref) => {
+    const { multiple, ...rest } = props;
+    const { files, createClient, ...fs } = useFileStore();
+    const { s3, ...storage } = useStorage();
+    const { credentials } = s3;
+
+    useEffect(() => {
+      async function init() {
+        useStorage.getState().initialize(api);
```
[the remark this answers is not in the document]
Comment (jamesacklin) [reply ui/src/components/FileUpload.tsx]: Calling directly now.

File: ui/src/components/FileUpload.tsx (hunk 3 above)
Comment (jamesacklin) [reply ui/src/components/FileUpload.tsx]: Cleaned this up a bit.

File: ui/src/components/FileUpload.tsx (hunk 2 above)
Comment (jamesacklin) [reply ui/src/components/FileUpload.tsx:54]: Cleaned this up, removed the `await`

File: ui/src/components/FileUpload.tsx (hunk 4 above)
Comment (jamesacklin) [reply ui/src/components/FileUpload.tsx:87]: Yes chef

**Review by patosullivan — APPROVED** [review]
LGTM!

**Pull request merged**